File: src/utils/WebSceneBridge.ts
import { RefObject } from 'react';
import { WebView } from 'react-native-webview';

export class WebSceneBridge {
  private webViewRef: RefObject<WebView> | null = null;
  private lastParallaxSentAt: number = 0;
  private readonly parallaxMinInterval: number = 1000 / 45; // ~45 fps

  constructor(webViewRef: RefObject<WebView> | null) {
    this.webViewRef = webViewRef;
  }

  update(webViewRef: RefObject<WebView> | null) {
    this.webViewRef = webViewRef;
  }

  getWebViewRef(): RefObject<WebView> | null {
    return this.webViewRef;
  }

  setCallMode(enabled: boolean) {
    const js = `window.setCallMode && window.setCallMode(${enabled ? 'true' : 'false'});`;
    this.evaluate(js);
  }

  applyParallax(dx: number, dy: number) {
    const now = Date.now();
    if (now - this.lastParallaxSentAt < this.parallaxMinInterval) {
      return;
    }
    this.lastParallaxSentAt = now;
    const js = `window.applyParallax && window.applyParallax(${dx}, ${dy});`;
    this.evaluate(js);
  }

  triggerDance() {
    const js = 'window.triggerDance && window.triggerDance();';
    this.evaluate(js);
  }

  stopAction() {
    // Stop any intense action (dance, etc) and return to idle
    const js = "window.loadAnimationByName && window.loadAnimationByName('Idle Stand');";
    this.evaluate(js);
  }

  loadAnimationByName(animName: string) {
    const safeName = animName.replace(/'/g, "\\'");
    const js = `window.loadAnimationByName && window.loadAnimationByName('${safeName}');`;
    this.evaluate(js);
  }

  playSpeech(text: string) {
    const duration = Math.min(Math.max(text.length * 60, 1500), 6000);
    const js = `
      (function(){
        if(window.setMouthOpen){
          window.setMouthOpen(0.6);
          setTimeout(() => window.setMouthOpen(0.05), ${duration});
        }
      })();
    `;
    this.evaluate(js);
  }

  private lastMouthLogTime = 0;

  setMouthOpen(value: number) {
    const clamped = Math.max(0, Math.min(1, Number.isFinite(value) ? value : 0));
    // Throttled debug log (max once per second)
    const now = Date.now();
    if (clamped > 0.1 && now - this.lastMouthLogTime > 1000) {
      this.lastMouthLogTime = now;
      console.log('[WebBridge] setMouthOpen:', clamped.toFixed(2), 'webViewRef:', !!this.webViewRef?.current);
    }
    const js = `window.setMouthOpen && window.setMouthOpen(${clamped.toFixed(3)});`;
    this.evaluate(js);
  }

  loadModel(js: string) {
    this.evaluate(js);
  }

  private evaluate(script: string) {
    if (this.webViewRef?.current) {
      this.webViewRef.current.injectJavaScript(script);
    }
  }
}

File: Roxie/Roxie/Components/WebSceneBridge.swift
import Foundation
import WebKit
import Combine

/// Full port of RN `src/utils/WebSceneBridge.ts`. All 9 RN→JS methods + incoming-message feed.
@MainActor
final class WebSceneBridge: ObservableObject {
    @Published private(set) var lastMessage: String?

    private weak var webView: WKWebView?
    private var pending: [() -> Void] = []

    private static let parallaxInterval: TimeInterval = 1.0 / 45.0
    private var lastParallaxAt: TimeInterval = 0
    private var lastMouthLogAt: TimeInterval = 0

    // MARK: - Lifecycle

    func attach(webView: WKWebView) {
        self.webView = webView
        let buffered = pending
        pending.removeAll()
        for command in buffered { command() }
    }

    func received(message: String) {
        lastMessage = message
    }

    // MARK: - Public bridge API

    func setCallMode(_ enabled: Bool) {
        evaluate("window.setCallMode && window.setCallMode(\(enabled ? "true" : "false"));")
    }

    func applyParallax(dx: Double, dy: Double) {
        let now = Date().timeIntervalSince1970
        guard now - lastParallaxAt >= Self.parallaxInterval else { return }
        lastParallaxAt = now
        evaluate("window.applyParallax && window.applyParallax(\(dx), \(dy));")
    }

    func triggerDance() {
        evaluate("window.triggerDance && window.triggerDance();")
    }

    func stopAction() {
        loadAnimation(named: "Idle Stand")
    }

    func loadAnimation(named animName: String) {
        let safe = animName.replacingOccurrences(of: "'", with: "\\'")
        evaluate("window.loadAnimationByName && window.loadAnimationByName('\(safe)');")
    }

    /// Intentionally a no-op during chat, matching RN. Voice mouth-sync happens via `setMouthOpen`.
    func playSpeech(_ text: String) { }

    func setMouthOpen(_ value: Double) {
        let clamped = max(0, min(1, value.isFinite ? value : 0))
        evaluate(String(format: "window.setMouthOpen && window.setMouthOpen(%.3f);", clamped))
    }

    /// Inject an arbitrary model-load JS (backend can ship model-specific loader snippets).
    func loadModel(_ js: String) {
        evaluate(js)
    }

    // MARK: - Private

    private func evaluate(_ js: String) {
        let run: () -> Void = { [weak self] in
            self?.webView?.evaluateJavaScript(js, completionHandler: nil)
        }
        if webView == nil {
            pending.append(run)
        } else {
            run()
        }
    }
}
